torch/csrc/dynamo/guards.cpp

```diff
@@ -464,10 +465,125 @@ static PyObject* assert_size_stride(PyObject* dummy, PyObject* args) {
   Py_RETURN_TRUE;
 }
 
+typedef struct {
+  PyDictObject *dict;  // borrowed reference
+  uint64_t dict_version_tag;
+} AttrTag;
+
+static const char* module_guard_attrs[] = {
+  "_parameters",
+  "_buffers",
+  "_modules",
+  "_forward_hooks",
+  "_forward_pre_hooks",
+};
+
+typedef struct {
+  PyObject_HEAD
+  PyObject *wr;
+  unsigned int version_tag;
+  uint64_t dict_version_tag;
+  AttrTag attr_tags[sizeof(module_guard_attrs) / sizeof(module_guard_attrs[0])];
+} NNModuleGuard;
+
+static void NNModuleGuard_dealloc(NNModuleGuard* self) {
+  Py_CLEAR(self->wr);
+  Py_TYPE(self)->tp_free((PyObject*)self);
+}
+
+static PyTypeObject NNModuleGuardType = {
+    // NOLINTNEXTLINE
+    PyVarObject_HEAD_INIT(NULL, 0)};
+
+static PyObject* NNModuleGuard_call(PyObject *callable, PyObject *args, PyObject *kwargs) {
+  NNModuleGuard *guard = (NNModuleGuard *)callable;
+
+  PyObject* mod = PyWeakref_GetObject(guard->wr);
+  if (mod == NULL) {
+    Py_RETURN_FALSE;
+  }
+
+  if (Py_TYPE(mod)->tp_version_tag != guard->version_tag) {
+    Py_RETURN_FALSE;
+  }
+
+  // NOTE: we must check the dict version tag before we check the attributes,
+  // because the attributes may be dead references if the dict has been updated.
+  PyObject* dict = PyObject_GenericGetDict(mod, NULL);
```

Comment:
Trying to understand. This will basically get `mod.__dict__` and will cover changes to member variables like `training`, right?

Reply:
i think this checks the version of mod.__dict__ in case, for instance, someone deleted `_parameters` from it and later on we would attempt to get the version of `mod.__dict__['_parameters']`

but, this check should also cover keys like `training` which being plain types would require changing the mod.__dict__ version if updating them?

(also just trying to check my understanding)

Reply:
Yes, any changes to attributes stored on the `__dict__` will lead to a change in `PyDictObject.ma_version_tag`.

For Python 3.12+, we'll want to use a different mechanism ([dictionary watchers](https://docs.python.org/3.12/c-api/dict.html#c.PyDict_AddWatcher)) because `ma_version_tag` is deprecated.

Reply:
@wconstab, yes I think your understanding is correct. If someone changes, something like `self.foo.bar = 3` that *won't* be caught it's mutating "foo" and not the module's `__dict__`.

Modifications to `_parameters`, `_buffers`, and hooks are handled in here: https://github.com/pytorch/pytorch/pull/100642/files#diff-f441b0c110772dd661a6b6cc9402f685eed03f0319597d91b661bf0bccd2d11aR473-R479.

Reply:
>  If someone changes, something like self.foo.bar = 3 that won't be caught it's mutating "foo" and not the module's __dict__

What about submodules? Like suppose we have larger nn.Module instance which instantiates many instances of smaller nn.Module instances in its `__init__` function. Do we need `NNModuleGuard` for each of these submodules?

Reply:
i think we'd need a guard on each nnmodule.  the handling on the root module's `_modules` dict would help notice any newly added or deleted submodule, but i don't think it would notice if anything inside that submodule got poked.